### PR TITLE
Restore timezone selection on GPDB startup

### PR DESCRIPTION
**Pull request by tvesely**
Commit b749790 backports changes from Postgres 10.1 to parse the most recent version of the IANA timezone library format, but these changes also included some binary incompatible changes in behavior. Specifically, `pg_timezone_pre_initialize()` and `identify_system_timezone()` were removed, resulting in the timezone defaulting to 'GMT' on database startup, even if the TimeZone GUC was specifically set in `postgresql.conf`.

This will restore the previous behavior and update time zone data files to tzdata release 2018c.

### CODE REVIEW

**Comment by yydzero** [comment]
@pivotal-mike @ivannovick we need get this into 5.8.

@magi345 for awareness

File: src/test/binary_swap/sql/timezone_check.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+-- If we properly detect the system timezone during startup, then the timezone
+-- will not be set to GMT.
+\! PGTZ='DEFAULT' psql binswap_connect -c "select count(*) from pg_settings where name='TimeZone' and setting != 'GMT';"
```
Comment (asimrp) [inline src/test/binary_swap/sql/timezone_check.sql]: Nice to see the test coverage.  What would the output be if this test runs on a system with timezone set to GMT?

File: src/backend/utils/misc/guc.c [hunk 2]
```diff
@@ -3357,12 +3357,17 @@ InitializeGUCOptions(void)
 	int			i;
 	char	   *env;
 	long		stack_rlimit;
+	/*
+	 * Before log_line_prefix could possibly receive a nonempty setting, make
+	 * sure that timezone processing is minimally alive (see elog.c).
+	 */
+	pg_timezone_pre_initialize();
```
Comment (asimrp) [inline src/backend/utils/misc/guc.c]: The call and the comment is redundant, let's remove them.

File: src/include/pgtime.h [hunk 3]
```diff
@@ -76,13 +76,7 @@ extern void pg_tzenumerate_end(pg_tzenum *dir);
...
- * GPDB_92_MERGE_FIXME: Remove when commit ca4af308c32 has been merged.
- * gmt_timezone has been made obsolete with log_timezone being set by
- * pg_timezone_initialize(), but in order to avoid merge conflicts in
- * the code the callers are still allowed to use gmt_timezone.
- */
-#define gmt_timezone log_timezone
+extern pg_tz *gmt_timezone;
 
 /*
  * GPDB_92_MERGE_FIXME: Remove when commit ca4af308c32 is merged. This
```
Comment (asimrp) [inline src/include/pgtime.h]: In 5X, there is no need to keep MERGE_FIXMEs.  Should we remove it?  Same question for [pgtz.c:21](https://github.com/greenplum-db/gpdb/pull/4896/commits/95b3c3f82f2b3989082b89cf996011e76b0a3cfd#diff-3f0b4f9d900b9d33e36470952f83f0a6R21)

File: src/test/binary_swap/sql/timezone_check.sql (hunk 1 above)
Comment (tvesely) [reply src/test/binary_swap/sql/timezone_check.sql]: The timezone would probably be set to GMT, and fail, but in practice nobody uses GMT, and rather use UTC if they want a generalized timezone.

This was the best compromise I could come up with, as we don't have timezone files installed in our test environment by default. If we ever run into a system with GMT as the default, I think we would need to reconsider this, however.

File: src/backend/utils/misc/guc.c (hunk 2 above)
Comment (tvesely) [reply src/backend/utils/misc/guc.c]: Good point. This double initialize must have been introduced in a merge with Postgres, and when I pulled the changes back from 5.4.1 it came along for the ride. I'll remove it.

File: src/include/pgtime.h (hunk 3 above)
Comment (tvesely) [reply src/include/pgtime.h]: Daniel's original commit pulled in all these merge FIXMEs. If I were to remove these ones, should I also be removing the half dozen others?

Comment (asimrp) [reply src/include/pgtime.h]: Ah, in that case, no harm in keeping them.

**Comment by pivotal-ning-yu** [comment]
Is there a plan to port commit https://github.com/greenplum-db/gpdb/pull/4896/commits/96262a08900dc9096b79a3b0e7fbdccea6b329eb, the binary swap tests, to master? Not sure if should we keep the binary swap tests the same on master and 5X branches.

**Comment by asimrp** [comment]
> Is there a plan to port commit 96262a0, the binary swap tests, to master?
Catalog is allowed to be changed on master branch.  Testing binary swap on master doesn't make sense.

**Comment by pivotal-ning-yu** [comment]
On Wed, May 9, 2018 at 1:00 AM, Asim RP <notifications@github.com> wrote:

> Is there a plan to port commit 96262a0
> <https://github.com/greenplum-db/gpdb/commit/96262a08900dc9096b79a3b0e7fbdccea6b329eb>,
> the binary swap tests, to master?
> Catalog is allowed to be changed on master branch. Testing binary swap on
> master doesn't make sense.
>
>
Yes, binary swap tests are only triggered on 5X pipeline, but as we have
put the binary swap tests in the master code branch should we also keep it
the same with 5X? Or shall we remove it entirely from master branch and
only keep and maintain it in 5X?

**Comment by ashwinstar** [comment]
On Tue, May 8, 2018 at 8:02 PM, Ning Yu <notifications@github.com> wrote:

> On Wed, May 9, 2018 at 1:00 AM, Asim RP <notifications@github.com> wrote:
>
> > Is there a plan to port commit 96262a0
> > <https://github.com/greenplum-db/gpdb/commit/
> 96262a08900dc9096b79a3b0e7fbdccea6b329eb>,
> > the binary swap tests, to master?
> > Catalog is allowed to be changed on master branch. Testing binary swap on
> > master doesn't make sense.
> >
> >
> Yes, binary swap tests are only triggered on 5X pipeline, but as we have
> put the binary swap tests in the master code branch should we also keep it
> the same with 5X? Or shall we remove it entirely from master branch and
> only keep and maintain it in 5X?
>

We should keep them in master branch as well I think, as will need them
once we freeze the catalog. It's little tricky situation though to have the
test code which we are not exercising and don't know if it works or not.
But seems still worth keeping the parity and when time to enable, only
broken pieces can be fixed vs attempting to find what's required.

**Pull request merged**